### PR TITLE
chore: migrate to typescript and jest

### PR DESCRIPTION
**Pull request by Nebulis**
- Migrated to typescript (removed rollup, babel)
- Migrated to jest (removed mocha, nyc, chai)
- Added @typescript-eslint

TODO:
- [ ] add types
- [ ] test builds

### CODE REVIEW

File: src/schema/schema.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+  }
+};
+
+export const validate = (document: Document, schema?: Schema) => {
+  // TODO document.schema is set as mandatory here because for the moment it can't be made required in the interface
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const result = schema ? ajv.validate(schema, getData(document)) : ajv.validate(document.schema!, getData(document));
+  // eslint-disable-next-line no-console
+  // console.log(ajv.errors); // TODO: properly feedback error
```
Comment (rjchow) [inline src/schema/schema.ts]: use node-debug mf

File: test/e2e.test.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
+import {
+  addSchema,
+  getData,
+  issueDocument,
+  issueDocuments,
+  obfuscateDocument,
+  validateSchema,
+  verifySignature
+} from "../src/index";
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply test/e2e.test.ts]: should run test in CI

File: src/privacy/privacy.ts [hunk 3]
```diff
@@ -1,11 +1,18 @@
-import { cloneDeep, pick, unset, get } from "lodash";
+import { cloneDeep, get, pick, unset } from "lodash";
 import { flatten } from "flat";
 import { toBuffer } from "../utils";
 import { unsaltData } from "./salt";
 
-export const getData = document => unsaltData(document.data);
+export interface Document {
+  data?: any;
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/privacy/privacy.ts]: need to have SignedDocument and UnsignedDocument

File: src/signature/signature.ts [hunk 4]
```diff
@@ -29,7 +28,7 @@ export const sign = (document, batch) => {
   });
 };
 
-export const verify = document => {
+export const verify = (document: Document) => {
   const signature = get(document, "signature");
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/signature/signature.ts]: proof is array of string

File: src/privacy/privacy.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/privacy/privacy.ts]: getData => SignedDocument
setData => ??
obfuscateDocument => SignedDocument and returns SignedDocument

File: src/signature/signature.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/signature/signature.ts]: verify can take any

File: src/signature/signature.ts [hunk 5]
```diff
@@ -42,7 +41,7 @@ export const verify = document => {
 
   // Calculates merkle root from target hash and proof, then compare to merkle root in document
   const merkleRoot = get(document, "signature.merkleRoot");
-  const proof = get(document, "signature.proof", []);
+  const proof: any[] = get(document, "signature.proof", []);
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/signature/signature.ts]: string[]

File: src/signature/merkle/merkle.ts [hunk 6]
```diff
@@ -47,11 +40,8 @@ function getLayers(elements) {
  * layer = [ A, B, C, D, E]
  * if index = 5, then return null
  * if index = 4, then return C
- *
- * @param {*} index
- * @param {*} layer
  */
-function getPair(index, layer) {
+function getPair(index: number, layer: any[]) {
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/signature/merkle/merkle.ts]: Buffer[]

File: src/signature/merkle/merkle.ts [hunk 7]
```diff
@@ -78,60 +65,48 @@ function getProof(index, layers) {
...
  * Function that runs through the supplied hashes to arrive at the supplied merkle root hash
- * @param {*} _proof The list of uncle hashes required to arrive at the supplied merkle root
- * @param {*} _root The merkle root
- * @param {*} _element The leaf node that is being verified
+ * @param _proof The list of uncle hashes required to arrive at the supplied merkle root
+ * @param _root The merkle root
+ * @param _element The leaf node that is being verified
  */
-export const checkProof = function(_proof, _root, _element) {
+export const checkProof = function(_proof: any[], _root: any, _element: any) {
```
[the remark this answers is not in the document]
Comment (yehjxraymond) [reply src/signature/merkle/merkle.ts]: hashToBuffer takes either a string or a Buffer
_proof is either a string or a Buffer
_proof is either a string or a Buffer
_element is either a string or a Buffer

File: test/e2e.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Nebulis) [reply test/e2e.test.ts]: let's do e2e later

**Pull request merged**